### PR TITLE
[FLINK-34877][cdc] Flink CDC pipeline transform supports type conversion

### PR DESCRIPTION
**Pull request by aiwenmo**
This closes [FLINK-34877](https://issues.apache.org/jira/browse/FLINK-34877).

### CODE REVIEW

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/parser/JaninoCompiler.java [hunk 1]
```diff
@@ -298,4 +312,33 @@ private static Java.Rvalue generateNoOperandTimestampFunctionOperation(String op
...
+            case "FLOAT":
+                return "convertFloat";
+            case "DOUBLE":
+                return "convertDouble";
+            case "DECIMAL":
+                return "convertBigDecimal";
+            case "CHAR":
+            case "VARCHAR":
+            case "STRING":
+            case "`STRING`":
```
Comment (yuxiqian) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/parser/JaninoCompiler.java:337]: Seems only the STRING type is specially treated to handle the quoted version. Will other types be quoted too? If so, maybe we can trim typename first before `switch` cases.

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/parser/JaninoCompiler.java [hunk 2]
```diff
@@ -298,4 +312,33 @@ private static Java.Rvalue generateNoOperandTimestampFunctionOperation(String op
...
+                return "convertString";
+            default:
+                throw new ParseException(
+                        "Unsupported data type convert: " + sqlDataTypeSpec.toString());
```
Comment (yuxiqian) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/parser/JaninoCompiler.java:341]:
```suggestion
                throw new ParseException(
                        "Unsupported data type cast: " + sqlDataTypeSpec.toString());
```

Using cast might be better since user writes `cast` instead of `convert` keyword in YAML

**Review by yuxiqian — COMMENTED** [review]
Thanks for @aiwenmo's great work! Just left some minor comments.

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/functions/SystemFunctionUtils.java [hunk 3]
```diff
@@ -510,4 +510,40 @@ public static Object coalesce(Object... objects) {
         }
         return null;
     }
+
+    public static String convertString(Object object) {
+        return object.toString();
```
Comment (lvyanquan) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/functions/SystemFunctionUtils.java:515]: What would happen if we pass a NULL Object to cast function? Does we allow this happen?

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/functions/SystemFunctionUtils.java [hunk 4]
```diff
@@ -510,4 +510,40 @@ public static Object coalesce(Object... objects) {
         }
         return null;
     }
+
+    public static String convertString(Object object) {
```
Comment (lvyanquan) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/functions/SystemFunctionUtils.java:514]: By the way, it's better to name them with `castToxxxx`.

**Review by lvyanquan — APPROVED** [review]
LGTM.

**Comment by aiwenmo** [comment]
@yuxiqian @leonardBang PTAL

**Comment by yuxiqian** [comment]
Is it possible to add actual values conversion checks besides existing syntax parsing tests? And it should cover cases where `NULL` value is passed as conversion argument.

**Comment by leonardBang** [comment]
> Is it possible to add actual values conversion checks besides existing syntax parsing tests? And it should cover cases where `NULL` value is passed as conversion argument.

@aiwenmo I think @yuxiqian's idea is reasonable, could you add more value tests ?

**Review by leonardBang — APPROVED** [review]
Thanks @aiwenmo for the update, LGTM

File: flink-cdc-runtime/src/test/java/org/apache/flink/cdc/runtime/operators/transform/TransformDataOperatorTest.java [hunk 5]
```diff
@@ -555,6 +574,64 @@ void testTimestampDiffTransform() throws Exception {
                 .isEqualTo(new StreamRecord<>(insertEventExpect));
     }
 
+    @Test
+    void testNullCastTransform() throws Exception {
```
Comment (yuxiqian) [inline flink-cdc-runtime/src/test/java/org/apache/flink/cdc/runtime/operators/transform/TransformDataOperatorTest.java:578]: Great to see this! Could you please insert more non-null data row to test 1) non-null value casting and verify results and 2) unsupported type conversion failure?

Comment (leonardBang) [reply flink-cdc-runtime/src/test/java/org/apache/flink/cdc/runtime/operators/transform/TransformDataOperatorTest.java:596]: @yuxiqian Looks like @aiwenmo is busy on other things, as this comment is a minor test-related improvement and this PR block some on-going PRs, I'll merge this PR soon.

**Review by leonardBang — APPROVED** [review]
+1 from my side

**Comment by aiwenmo** [comment]
Hi. I will commit the test tonight. And the code has been completed.

**Comment by leonardBang** [comment]
> Hi. I will commit the test tonight. And the code has been completed.

Thanks @aiwenmo ，I'll wait your test code improvement in this PR

**Review by yuxiqian — APPROVED** [review]
Thanks for @aiwenmo's great work! @leonardBang PTAL
